File: src/main/java/net/gegy1000/terrarium/server/map/GenerationRegionHandler.java
package net.gegy1000.terrarium.server.map;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.collect.Lists;
import net.gegy1000.terrarium.Terrarium;
import net.gegy1000.terrarium.server.capability.TerrariumWorldData;
import net.gegy1000.terrarium.server.map.adapter.CoastlineAdapter;
import net.gegy1000.terrarium.server.map.adapter.WaterFlattenAdapter;
import net.gegy1000.terrarium.server.map.source.osm.OverpassTileAccess;
import net.gegy1000.terrarium.server.map.system.RegionPopulationSystem;
import net.gegy1000.terrarium.server.map.system.component.TerrariumComponentTypes;
import net.gegy1000.terrarium.server.map.system.populator.CoverRegionPopulator;
import net.gegy1000.terrarium.server.map.system.populator.HeightRegionPopulator;
import net.gegy1000.terrarium.server.map.system.populator.OverpassRegionPopulator;
import net.gegy1000.terrarium.server.map.system.populator.SlopeRegionPopulator;
import net.gegy1000.terrarium.server.map.system.sampler.DataSampler;
import net.gegy1000.terrarium.server.map.system.sampler.GlobSampler;
import net.gegy1000.terrarium.server.map.system.sampler.HeightSampler;
import net.gegy1000.terrarium.server.map.system.sampler.OverpassSampler;
import net.gegy1000.terrarium.server.map.system.sampler.SlopeSampler;
import net.gegy1000.terrarium.server.util.Coordinate;
import net.gegy1000.terrarium.server.util.Interpolation;
import net.gegy1000.terrarium.server.world.EarthGenerationSettings;
import net.minecraft.world.World;

import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class GenerationRegionHandler {
    private final EarthGenerationSettings settings;
    private final TerrariumWorldData worldData;

    private final Coordinate bufferedRegionSize;

    private final LoadingCache<RegionTilePos, GenerationRegion> cache = CacheBuilder.newBuilder()
            .expireAfterAccess(30, TimeUnit.SECONDS)
            .maximumSize(8)
            .build(new CacheLoader<RegionTilePos, GenerationRegion>() {
                @Override
                public GenerationRegion load(RegionTilePos key) {
                    try {
                        return GenerationRegionHandler.this.generate(key);
                    } catch (Exception e) {
                        Terrarium.LOGGER.error("Failed to load generation region at {}", key, e);
                    }
                    return GenerationRegionHandler.this.createDefaultRegion(key);
                }
            });

    private final RegionPopulationSystem populationSystem;

    public GenerationRegionHandler(World world, EarthGenerationSettings settings, TerrariumWorldData worldData) {
        this.settings = settings;
        this.worldData = worldData;

        this.bufferedRegionSize = Coordinate.fromBlock(settings, GenerationRegion.BUFFERED_SIZE, GenerationRegion.BUFFERED_SIZE);

        HeightSampler heightSampler = new HeightSampler(worldData.getHeightSource());
        SlopeSampler slopeSampler = new SlopeSampler(heightSampler);
        GlobSampler coverSampler = new GlobSampler(worldData.getGlobSource());
        List<DataSampler<OverpassTileAccess>> overpassSamplers = Lists.newArrayList(
                new OverpassSampler(worldData.getOutlineOverpassSource()),
                new OverpassSampler(worldData.getGeneralOverpassSource()),
                new OverpassSampler(worldData.getDetailedOverpassSource())
        );

        Interpolation.Method interpolationMethod = Interpolation.Method.CUBIC;
        double scale = 1.0 / settings.worldScale;
        if (scale >= 45.0) {
            interpolationMethod = Interpolation.Method.LINEAR;
        } else if (scale >= 20.0) {
            interpolationMethod = Interpolation.Method.COSINE;
        }

        Random random = new Random(world.getSeed());
        this.populationSystem = RegionPopulationSystem.builder(settings)
                .withComponent(TerrariumComponentTypes.HEIGHT, new HeightRegionPopulator(random, heightSampler, interpolationMethod))
                .withComponent(TerrariumComponentTypes.SLOPE, new SlopeRegionPopulator(slopeSampler))
                .withComponent(TerrariumComponentTypes.COVER, new CoverRegionPopulator(coverSampler))
                .withComponent(TerrariumComponentTypes.OVERPASS, new OverpassRegionPopulator(overpassSamplers))
                .withAdapter(new CoastlineAdapter())
                .withAdapter(new WaterFlattenAdapter(16))
                .build();
    }

    public GenerationRegion get(int blockX, int blockZ) {
        return this.get(new RegionTilePos(Math.floorDiv(blockX, GenerationRegion.SIZE), Math.floorDiv(blockZ, GenerationRegion.SIZE)));
    }

    public GenerationRegion get(RegionTilePos pos) {
        try {
            return this.cache.get(pos);
        } catch (ExecutionException e) {
            Terrarium.LOGGER.error("Failed to retrieve generation region from cache at {}", pos, e);
        }
        return this.createDefaultRegion(pos);
    }

    private GenerationRegion generate(RegionTilePos pos) {
        Coordinate minCoordinate = pos.getMinBufferedCoordinate(this.settings);
        Coordinate maxCoordinate = pos.getMaxBufferedCoordinate(this.settings);

        if (!minCoordinate.inWorldBounds() || !maxCoordinate.inWorldBounds()) {
            return this.createDefaultRegion(pos);
        }

        RegionData data = this.populationSystem.populateData(pos, this.bufferedRegionSize, GenerationRegion.BUFFERED_SIZE, GenerationRegion.BUFFERED_SIZE);
        return new GenerationRegion(pos, data);
    }

    private GenerationRegion createDefaultRegion(RegionTilePos pos) {
        return new GenerationRegion(pos, new RegionData(Collections.emptyMap()));
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/GenerationRegionHandler.kt
package net.gegy1000.terrarium.server.map

import com.google.common.cache.CacheBuilder
import com.google.common.cache.CacheLoader
import net.gegy1000.terrarium.server.capability.TerrariumWorldData
import net.gegy1000.terrarium.server.map.adapter.CoastlineAdapter
import net.gegy1000.terrarium.server.map.glob.GlobType
import net.gegy1000.terrarium.server.map.source.osm.OverpassTileAccess
import net.gegy1000.terrarium.server.util.Coordinate
import net.gegy1000.terrarium.server.world.generator.EarthGenerationHandler
import net.minecraft.util.math.MathHelper
import java.util.concurrent.TimeUnit

class GenerationRegionHandler(val worldData: TerrariumWorldData, val generator: EarthGenerationHandler) {
    companion object {
        val adapters = arrayOf(CoastlineAdapter)
    }

    private val cache = CacheBuilder.newBuilder()
            .expireAfterAccess(30, TimeUnit.SECONDS)
            .maximumSize(9)
            .build(object : CacheLoader<RegionTilePos, GenerationRegion>() {
                override fun load(pos: RegionTilePos) = generate(pos)
            })

    private val scaledDataSize = MathHelper.floor(GenerationRegion.SAMPLE_SIZE * this.generator.finalScale)

    private val regionSampleSize = GenerationRegion.SAMPLE_SIZE + 1

    private val sampledHeights = ShortArray(this.regionSampleSize * this.regionSampleSize)
    private val sampledGlobs = Array(this.regionSampleSize * this.regionSampleSize, { GlobType.NO_DATA })

    operator fun get(globalX: Int, globalZ: Int): GenerationRegion {
        val pos = RegionTilePos(MathHelper.floor(globalX.toDouble() / GenerationRegion.SIZE), MathHelper.floor(globalZ.toDouble() / GenerationRegion.SIZE))
        return this[pos]
    }

    operator fun get(position: RegionTilePos) = this.cache[position]!!

    private fun generate(position: RegionTilePos): GenerationRegion {
        this.generator.initializeSeed(position)

        val minCoordinate = position.getMinCoordinate(generator.settings)
        val maxCoordinate = position.getMaxCoordinate(generator.settings).addGlobal(2.0, 2.0)
        val overpassTile = worldData.overpassSource.sampleArea(minCoordinate, maxCoordinate)

        val resultHeights = this.generateHeights(minCoordinate, maxCoordinate)
        val resultGlobs = this.generateGlobcover(overpassTile, minCoordinate, maxCoordinate)

        return GenerationRegion(position, minCoordinate, scaledDataSize, resultHeights, resultGlobs)
    }

    private fun generateHeights(minCoordinate: Coordinate, maxCoordinate: Coordinate): ShortArray {
        worldData.heightSource.sampleArea(sampledHeights, minCoordinate, maxCoordinate)

        val resultHeights = ShortArray(scaledDataSize * scaledDataSize)
        generator.scaleHeightRegion(resultHeights, sampledHeights, GenerationRegion.SAMPLE_SIZE + 1, GenerationRegion.SAMPLE_SIZE + 1, scaledDataSize, scaledDataSize)

        return resultHeights
    }

    private fun generateGlobcover(overpassTile: OverpassTileAccess, minCoordinate: Coordinate, maxCoordinate: Coordinate): Array<GlobType> {
        worldData.globSource.sampleArea(sampledGlobs, minCoordinate, maxCoordinate)

        val resultGlobs = Array(scaledDataSize * scaledDataSize, { GlobType.NO_DATA })
        generator.scaleGlobRegion(resultGlobs, sampledGlobs, GenerationRegion.SAMPLE_SIZE + 1, GenerationRegion.SAMPLE_SIZE + 1, scaledDataSize, scaledDataSize)

        val originX = MathHelper.floor(minCoordinate.blockX)
        val originZ = MathHelper.floor(minCoordinate.blockZ)
        adapters.forEach { adapter ->
            adapter.adaptGlobcover(generator.settings, overpassTile, resultGlobs, originX, originZ, scaledDataSize, scaledDataSize)
        }

        return resultGlobs
    }
}
